src/Physics/Multinucleon/XSection/MECScaleVsW.cxx

```diff
@@ -0,0 +1,170 @@
+  double Q3 = interaction.Kine().GetKV(kKVQ3) ; 
+
+  return GetScaling( Q0, Q3 ) ; 
+}
+//_________________________________________________________________________
+double MECScaleVsW::GetScaling( const double Q0, const double Q3 ) const
+{
+  // Get the vectors that include the kinematic limits of W for a given event:
+  MECScaleVsW::weight_type_map weight_map = GetMapWithLimits( Q0, Q3 ) ;
+  
+  // The Scaling is done using the "experimenter's W", which assumes a single nucleon
+  // See motivation in : https://arxiv.org/pdf/1601.02038.pdf
+  // Calculate event W:
+  static double Mn = ( PDGLibrary::Instance()->Find(kPdgProton)->Mass() + PDGLibrary::Instance()->Find(kPdgNeutron)->Mass() ) * 0.5 ;  // Nucleon mass
+  double W = sqrt( pow(Mn,2) + 2*Mn*Q0 - pow(Q3,2) + pow(Q0,2) ) ;
+
+  // Calculate scaling:
+  unsigned int step = ( weight_map.size() -1 ) / 2 ;
+  MECScaleVsW::weight_type_map::iterator it = std::next( weight_map.begin(), step ) ; 
+  
+  while ( step < weight_map.size() && step > 1 ) {
+    MECScaleVsW::weight_type_map::iterator it_next = std::next( it ) ; 
+    if( W > it -> first ) {
+      if( W < it_next -> first ) {
+	return ScaleFunction( W, *it, *it_next ) ;  
+      } else { 
+	step += step / 2 ; 
+	it = std::next( weight_map.begin(), step  ) ; 
+      }
+    } else {
+      step -= step / 2 ; 
+      it = std::next( weight_map.begin(), step  ) ; 
+    }
+  }
+
+  return 1. ; 
+} 
+
+//_________________________________________________________________________
+MECScaleVsW::weight_type_map MECScaleVsW::GetMapWithLimits( double Q0, double Q3 ) const {
+  // This function is responsible to add the phase space limits in the WValues vector in case they are not included
+  // in the configuration setup.
+
+  static double Mn = ( PDGLibrary::Instance()->Find(kPdgProton)->Mass() + PDGLibrary::Instance()->Find(kPdgNeutron)->Mass() ) * 0.5 ;  // Nucleon mass
+  double W_min = sqrt( pow(Mn,2) + 2*Mn*fW1_Q0Q3_limits.Eval(Q3) - pow(Q3,2) + pow(fW1_Q0Q3_limits.Eval(Q3),2) ) ;
+  double W_max = sqrt( pow(Mn,2) + 2*Mn*Q0 ) ; // Imposing Q2 = 0 
+
+  // Insert phase space limits:
+  MECScaleVsW::weight_type_map w_map = fWeightsMap ; 
+  w_map.insert( weight_type_pair( W_max, fDefaultWeight ) ) ;
+  w_map.insert( weight_type_pair( W_min, fDefaultWeight ) ) ;
+
+  return w_map ; 
+} 
+//_________________________________________________________________________
+    
+double MECScaleVsW::ScaleFunction( double W, weight_type_pair min, weight_type_pair max ) const 
+{
+  // This function is responsible to calculate the scale at a given W
+  // It interpolates the value between scale_min (W_min) and scale_max (W_max) linearly 
+  // Get Wpoints from pair:
+  double W_min = min.first ; 
+  double W_max = max.first ; 
+  // Get scale coeficients at W_min and W_max:
+  double scale_min = min.second ; 
+  double scale_max = max.second ; 
+
+  double scale = ( scale_min - scale_max ) * ( W - W_min ) / ( W_min - W_max ) + scale_min ; 
```

Comment:
Just use min and max, and please use consistent notation to avoid heart attacks in the reader